daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetInputter.scala

```diff
@@ -400,114 +427,283 @@ abstract class InfosetInputter
-// NOTE: for large blob/clob objects there's a reason for a start/end event
-// because we may want to open a stream to get the contents in between without
-// allocating a string to hold it.
+// In Daffodil, the two events will always be back-to-back.
 //
-sealed trait InfosetEventKind
-case object StartKind extends InfosetEventKind { override def toString = "start" }
-case object EndKind extends InfosetEventKind { override def toString = "end" }
+sealed trait InfosetEventKind {
+
+  /**
+   * Converts name like StartArray into "array start"
+   *
+   * Error messages depend on this word order.
+   */
+  override def toString() = {
+    val cn = Misc.getNameFromClass(this)
+    val initialLC = Misc.initialLowerCase(cn)
+    val (se, ea) = initialLC.span(_.isLower)
+    val res = ea.toLowerCase() + " " + se
+    res
+  }
+}
+
+/**
+ * InfosetEvents are of these kinds.
+ */
+object InfosetEventKind {
+  sealed trait StartKind extends InfosetEventKind
+  sealed trait EndKind extends InfosetEventKind
+  sealed trait ArrayKind extends InfosetEventKind
+  sealed trait ElementKind extends InfosetEventKind
+  case object StartElement extends StartKind with ElementKind
+  case object EndElement extends EndKind with ElementKind
+  case object StartArray extends StartKind with ArrayKind
+  case object EndArray extends EndKind with ArrayKind
+}
 
 /**
  * An infoset event accessor.
  *
  * Not a case class because we don't want pattern matching on this. (It allocates to pattern match)
  *
- * We have erd as a seperate field (instead of just using node.erd) since
- * the inspectPure method finds an erd without constructing the corresponding DINode (so as to avoid side effects)
+ * The info associated with an infoset event is either a DIElement (for elements), or
+ * an ElementRuntimeData for arrays. We don't want to use a DIArray object because those
+ * are the private managed storage of DIComplex elements.
+ *
+ * At the time these info DIElements are created to carry the infoset event information
+ * they are not attached to parents/children in the infoset.
+ *
+ * We are effectively just using them as event "tuples" to carry the ERD and simple value
+ * information after having resolved the incoming event's local name + namespace (if we're using those)
+ * to get the ERD.
+ *
+ * By using our DIElement objects as the "tuples" of event information, we avoid having to
+ * allocate yet another separate object for these.
  *
- * We maintaint the invariant that node==null || node.erd==erd
+ * The Unparsers will examine the incoming DIElement, and associate them parent with child if
+ * they decide to keep them.
  *
+ * Unparsers are also responsible for looking at the incoming event, and deciding if we
+ * should first fill in any defaultable elements, or compute any dfdl:outputValueCalc elements.
+ *
+ * This changes the behavior of infoset inputters. You can no longer just pull events and have them
+ * get connected together into an infoset.
```

Comment:
Phrasing. Comments shouldn't describe behavior in terms of previous versions of the code.